juju/juju
# Add uniter support for running the secret-changed hook

This PR adds the support for running the secret-changed hook. Two new secrets manager api calls are added to allow the uniter to watch for secret changes and to ask for the most recent revision info.
The secrets resolved and local/remote state is updated to manage the known revision info.

## Checklist


## QA steps

bootstrap and deploy the ubuntu charm
create a secret, eg
```
juju exec --unit ubuntu/0 "secret-add foo-bar
secret-get secret:cbqqf59esa1idj8gu59g
```

deploy a second charm with a secret-changed hook, eg
```
$ cat hooks/secret-changed 
#!/bin/bash
echo "secret-changed"
juju-log "secret-changed uri=$JUJU_SECRET_URI label=$JUJU_SECRET_LABEL"
exit 0
```

get the secret value
```
juju exec --unit consumer/0 "secret-get secret:cbqqf59esa1idj8gu59g --label baz"
foo=bar
```

check the logs - no secret changed hook
update the secret
`juju exec --unit ubuntu/0 "secret-update secret:cbqqf59esa1idj8gu59g foo=bar2"`

check the logs
```
INFO juju.worker.uniter.operation ran "secret-changed" hook (via explicit, bespoke hook script)
INFO unit.consumer/0.juju-log secret-changed uri=secret:cbqqf59esa1idj8gu59g label=baz
```
